### PR TITLE
Design and simulation of PCR

### PR DESCRIPTION
**Pull request by Koeng101**
I need to go over docs for this code + more tests, but generally, this PCR functionality seems to work. 

I've added 3 functions in the PCR subpackage within /primers/:

- DesignPrimersWithOverhang
- DesignPrimers
- Simulate

These should cover the basic use cases for PCRing.

### CODE REVIEW

**Comment by TimothyStiles** [comment]
@Koeng101 reviewing now and got rid of the silly code climate issues.

The six things left brought up from code climate are legit and should be fixed. Will review and add comments.

File: primers/pcr/pcr.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	forwardPrimer = forwardOverhang + forwardPrimer
+	reversePrimer = transform.ReverseComplement(reverseOverhang) + reversePrimer
+
+	return forwardPrimer, reversePrimer
+}
+
+func DesignPrimers(sequence string, targetTm float64) (string, string) {
+	return DesignPrimersWithOverhangs(sequence, "", "", targetTm)
+}
+
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:35]: Needs comment for docs.

File: primers/pcr/pcr.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		reversePrimer = transform.ReverseComplement(sequence[len(sequence)-(minimalPrimerLength+i):])
+	}
+
+	// Add overhangs to primer
+	forwardPrimer = forwardOverhang + forwardPrimer
+	reversePrimer = transform.ReverseComplement(reverseOverhang) + reversePrimer
+
+	return forwardPrimer, reversePrimer
+}
+
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:31]: Needs comment for docs.

File: primers/pcr/pcr.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/TimothyStiles/poly/transform"
+	"index/suffixarray"
+	"sort"
+	"strings"
+)
+
+// https://doi.org/10.1089/dna.1994.13.75
+var minimalPrimerLength int = 15
+
+func DesignPrimersWithOverhangs(sequence, forwardOverhang, reverseOverhang string, targetTm float64) (string, string) {
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:14]: Needs comment for docs.

Comment (TimothyStiles) [reply primers/pcr/pcr.go:14]: Can you also give some overview on how primer design works? What are overhangs and why are they needed?, etc.

File: primers/pcr/pcr.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+	var pcrFragment string
+	var pcrFragments []string
+
+	// Next, iterate through the forwardLocations list
+	for i, forwardLocation := range forwardLocationInts {
+		// First, make sure that this isn't the last element in forwardLocations
+		if i+1 != len(forwardLocationInts) {
+			// If this isn't the last element in forwardLocations, then we can select the first reverseLocation that is less than the next forwardLocation
+			for _, reverseLocation := range reverseLocationInts {
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:85]: This block seems to be repeated three times in this loop based on very minimal conditionals. Can this be refactored into a helper function for better code clarity?

File: primers/pcr/pcr.go (hunk 1 above)
Comment (TimothyStiles) [reply primers/pcr/pcr.go:35]: This also needs more in-line context. Regular devs won't know anything about forward and reverse locations, minimal length, melting temp, overhangs or even what PCR is.

File: primers/pcr/pcr.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	return DesignPrimersWithOverhangs(sequence, "", "", targetTm)
+}
+
+func Simulate(sequence string, targetTm float64, circular bool, primerList []string) []string {
+	sequence = strings.ToUpper(sequence)
+	sequenceIndex := suffixarray.New([]byte(sequence))
+
+	primerLength := len(primerList)
+
+	forwardLocations := make(map[int][]int)
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:42]: explain what this is.

File: primers/pcr/pcr.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	reverseLocations := make(map[int][]int)
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:43]: explain what this is.

File: primers/pcr/pcr.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	minimalPrimers := make([]string, primerLength)
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:44]: What are minimal primers here?

File: primers/pcr/pcr.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	for primerIdx, primer := range primerList {
+		var minimalLength int
+		for i := 10; primers.MeltingTemp(primer[len(primer)-i:]) < targetTm; i++ {
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:47]: magic number. Why 10?

File: primers/pcr/pcr.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+			minimalLength = i
+			if primer[len(primer)-i:] == primer {
+				break
+			}
+		}
+		// Use the minimal binding sites of the primer to find positions in the genome
```
Comment (TimothyStiles) [inline primers/pcr/pcr.go:53]: What's a minimal binding site?

**Review by TimothyStiles — CHANGES_REQUESTED** [review]
Code quality overall looks good for the most part. There's a complex repeated block that should be turned into a helper function and overall this needs a lot more biological context before it's merged.

**Comment by isaacguerreir** [comment]
Had added a PR to add more functions for this subpackage #188

**Comment by Koeng101** [comment]
I am closing this PR because I think a few things need to be changed with the PCR simulation (for example, multiple templates at once for multiplexing). These changes are non-trivial and there are other PRs to focus on right now. I will add it back when possible.

**Pull request closed without merging**